In [ ]:
# ==========================================
# Modelo Supervisado con KNN
# Dataset público: Wine Dataset
# ==========================================

# 1. Importar librerías
import pandas as pd

from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [ ]:
# 2. Cargar el dataset
wine = load_wine()

X = wine.data          # Variables predictoras
y = wine.target        # Variable objetivo: tipo de vino

El dataset de vino contiene propiedades químicas de diferentes vinos, como el alcohol, el ácido málico, la ceniza, etc.
La etiqueta (variable `y`) representa el tipo o clase de vino, indicando su origen o variedad.
Permite clasificar los vinos basándose en sus características analíticas.

In [ ]:
# 3. Convertir a DataFrame solo para visualizar mejor
df = pd.DataFrame(X, columns=wine.feature_names)
df["target"] = y

print("Primeras filas del dataset:")
print(df.head())

Primeras filas del dataset:
   alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  p

In [ ]:
# 4. Dividir los datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.20,
    random_state=42
)


In [ ]:
# 5. Estandarizar los datos
# KNN usa distancias, por eso es importante que las variables estén en la misma escala
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)   # Aprende la escala con datos de entrenamiento
X_test_scaled = scaler.transform(X_test)         # Aplica la misma escala a los datos de prueba


In [ ]:
# 6. Crear el modelo KNN, con k=5
model = KNeighborsClassifier(n_neighbors=5)

In [ ]:
# 7. Entrenar el modelo
model.fit(X_train_scaled, y_train)

KNeighborsClassifier()

In [ ]:
# 8. Hacer predicciones
y_pred = model.predict(X_test_scaled)

In [ ]:
# 9. Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)

print("\nAccuracy del modelo:")
print(accuracy)

print("\nMatriz de confusión:")
print(confusion_matrix(y_test, y_pred))

print("\nReporte de clasificación:")
print(classification_report(y_test, y_pred))


Accuracy del modelo:
0.9444444444444444

Matriz de confusión:
[[14  0  0]
 [ 1 12  1]
 [ 0  0  8]]

Reporte de clasificación:
              precision    recall  f1-score   support

           0       0.93      1.00      0.97        14
           1       1.00      0.86      0.92        14
           2       0.89      1.00      0.94         8

    accuracy                           0.94        36
   macro avg       0.94      0.95      0.94        36
weighted avg       0.95      0.94      0.94        36



El modelo KNN obtuvo una **precisión general del 94.4%**, lo que indica que clasificó correctamente la mayoría de los vinos.
La **matriz de confusión** muestra un rendimiento excelente para las clases 0 y 2, con algunas confusiones mínimas entre la clase 1 y la clase 2.
El **reporte de clasificación** detalla que el modelo es robusto para todas las clases, con alta precisión y recall, especialmente para las clases 0 y 2.

In [ ]:
# 10. Usar el modelo con un nuevo caso
nuevo_vino = [X_test[5]]  # Tomamos un ejemplo del set de prueba

# Escalamos la data
nuevo_vino_scaled = scaler.transform(nuevo_vino)

prediccion = model.predict(nuevo_vino_scaled)

print("\nPredicción para un nuevo vino:")
print("Clase predicha:", prediccion[0])


Predicción para un nuevo vino:
Clase predicha: 0
